File: 173-二叉搜索树迭代器.py
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.mystack = []
        self.current = root
        while self.current:
            self.mystack.append(self.current)
            self.current = self.current.left

    def next(self) -> int:
        """
        @return the next smallest number
        """
        result = self.mystack.pop()
        self.current = result.right
        while self.current:
            self.mystack.append(self.current)
            self.current = self.current.left
        return result.val
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.mystack != []

File: 173-二叉搜索树迭代器.cpp
class BSTIterator {
public:
    stack<TreeNode*> mystack;
    TreeNode* current;

    BSTIterator(TreeNode* root) {
        current = root;
        while(current){
            mystack.push(current);
            current = current->left;
        }
    }
    
    /** @return the next smallest number */
    int next() {
        current = mystack.top();
        int result = current->val;
        current = current->right;
        mystack.pop();
        
        while(current){
            mystack.push(current);
            current = current->left;
        }
        
        return result;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !mystack.empty();
    }
};
